server/controllers/contest.js

```diff
@@ -30,7 +27,62 @@ exports.putWinner = async (req, res, next) => {
     if (!contest) {
       return res.status(404).json({ msg: "contest not found" });
     }
-    console.log(contest);
+    const { users, io } = req.app.io;
+    const notification = new Notification({
+      message:
+        "congratulations you have been chosen as the winner of the contest!",
+      priority: "high",
+      read: false,
+      notifOwner: creator,
+      link: `/contest/${contestId}`
+    });
+
+    await notification.save();
+    if (users[creator._id]) {
+      io.in(users[creator._id]).emit("addNotification", {
+        notification
+      });
+    }
+
+    const notificationsArray = [];
+    for (let i = 0; i < contest.submissions.length; i++) {
+      const endedNotification = new Notification({
+        message: "the contest has ended",
+        priority: "medium",
+        read: false,
+        notifOwner: contest.submissions[i].creator._id,
+        link: `/contest/${contestId}`
+      });
+      notificationsArray.push(endedNotification);
+    }
+
+    const unique = notificationsArray.reduce(
```

Comment:
this seems unnecessary if submissions route denied extra submissions like it's supposed to. Since it doesn't, I feel like this could still be written better. Maybe:
const uniqueIds = [];
const unique = notificationsArray.filter(notif => {
....const result = !uniqueIds.includes(notif.notifOwner);  // if it's not in unique return true
....if (result) uniqueIds.push(notif.notifOwner);                // push it in
....return result;                                                                // returning true keeps the notification
}); // early periods indicate tab
Regardless, I dislike the usage of a and b as variables in reduce, as it's not a comparison like sort. The two are not the same thing (one is an object with temp array and out array, the other is a notification object).